stackrox/stackrox
# chore(ui): Encapsulate entity nouns for Compliance

## Description

Counterpart to #6753

### Problems

Removal of excess style for accessibility raised the profile of inconsistencies in case and plurality of entity nouns.

Ironically, global entity types and names:
* hurt instead of help consistent case and plural because instead of solving in one place, there are many lodash function calls to adjust particular occurrences
* are difficult to delete when containers become obsolete

### Analysis

Find in Files `resourceLabels` has 32 results in 15 files

| folder | results | files |
| :--- | ---: | ---: |
| src/Components | 4 | 2 |
| src/Containers/AppPageTitle.tsx | 2 | 1 |
| src/Containers/Compliance | 21 | 9 |
| src/Containers/VulnManagement | 2 | 1 |
| src/messages | 3 | 2 |
| | 32 | 15 |

Here are pages that render each edited component:
1. Dashboard/ComplianceDashboardTile.tsx
    * /main/compliance: 4 link buttons
2. Entity/Header.js
    * /main/compliance/whatever/id: **Whatever** in page subheading.
3. List/Table.js
    * /main/compliance and click **View Standard** in sunburst widget: **n controls across n categories** in table heading.
    * /main/compliance/whatevers: **n whatevers** in table heading.
    * /main/compliance/whatever/id and click an entities tab: **n whatevers** in table heading.
    * /main/compliance/whatever/id and click Controls tab: **n controls across n standards** in table heading.
4. widgets/ControlRelatedResourceList.js
    * /main/compliance/controls/id sometimes see **n related whatevers** widget in side panels
5. widgets/EntityCompliance.js
    * /main/compliance/whatevers/id: **Whatever Compliance** in side panel.
6. widgets/ResourceCount.js
    * /main/compliance/clusters/id scroll down to 3 Whatever Count widgets
    * /main/compliance/namespaces/id scroll down to 2 Whatever Count widgets
7. widgets/ResourceRelatedResourceList.js is obsolete so delete
    * added on 2019-02-08 in rox 1283
    * superseded but not delete on 2019-03-26 in rox 1557
8. widgets/StandardsAcrossEntity.js
    * /main/compliance: Passing standards across whatevers
    * Bonus delete `id` prop of widget and replace `data-testid` attribute in tests.
9. widgets/StandardsByEntity.js:
    * /main/compliance: Passing standards by cluster

### Solution

Add entitiesForCompliance.ts as source of truth for entity names to replace occurrences of global `resourceLabels` in the container.

### Lessons

1. At least some of components in widgets folder are imported in only one of Dashboard, Entity, List folders. Encapsulating not just within container, but within subfolder might make less likely that a widget is obsolete.
2. Similarly, explicit routes instead of implicity page map might make less likely that Standard page might have become obsolete.
3. Although Compliance preceded entity relationships for Configuration Management and Vulnerability Management, entities like SECRET that have limited specific relationships do not fit a generic model. Similarly for entities like IMAGE and NODE that have limited specific relationships to IMAGE_COMPONENT, IMAGE_CVE, and so on.

## Checklist

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui
3. `yarn start` in ui

### Manual testing

The subtlety of **across n whatever** tripped me up in the recent removal of uppercase style for accessibility. Even this second time, it fried my brain.

1. Visit /main/compliance and scan
    * See count and noun in 4 dashboard tile links.
    * See plural noun in heading of 4 widgets.
    ![dashboard_after_scan](https://github.com/stackrox/stackrox/assets/11862657/58f7506b-91ed-4135-bd5f-53e0b5449ac8)

2. Click a standard link in a **Passing standards** widget
    * See **across 1 cluster** in table heading
    ![controls](https://github.com/stackrox/stackrox/assets/11862657/69c38dde-5e9a-495f-a466-df88663d710e)

3. Click a control
    * See related entities widgets (in some but not all)
    ![controls_id](https://github.com/stackrox/stackrox/assets/11862657/ecc2a7e7-bb80-4116-a67f-1de1c98dfa9b)

4. Click a **View Standard** link in a sunburst widget
    * See **across n categories** in table heading.
    ![controls_category](https://github.com/stackrox/stackrox/assets/11862657/f5bc2f43-0d31-49b5-be7a-909cfdeada75)

5. Visit /main/compliance/clusters for example
    * See plural noun in page heading.
    * See count and noun in table heading.
    ![clusters](https://github.com/stackrox/stackrox/assets/11862657/93d3b1f3-ed7c-4e51-9901-eab42a99b15d)

6. Click a row to open side panel
    * See **Cluster Compliance** widget at beginning.
    * See **Whatever Count** widgets at end.
    ![clusters_id_1](https://github.com/stackrox/stackrox/assets/11862657/51002aca-3a13-410b-9bb4-1ffb81489612)
    ![clusters_id_2](https://github.com/stackrox/stackrox/assets/11862657/bbd43f50-9015-4c36-954f-eb476f460150)

7. Click link to single page
    * See singular noun in page subheading.
    * See **Cluster Compliance** widget on **Overview** tab.
    * Click **controls** tab and see **across n standards** in table heading.
    * Click an entity tab and see table heading.
    ![cluster_id_1](https://github.com/stackrox/stackrox/assets/11862657/601f6b8d-1995-4ad7-9f55-d143e0a31e60)
    ![cluster_id_2](https://github.com/stackrox/stackrox/assets/11862657/80b2f975-f011-475e-ba23-7164fca03fe6)
    ![cluster_id_3](https://github.com/stackrox/stackrox/assets/11862657/88cff608-63c3-4cbd-8998-d80ab8f68ff0)

### Integration testing

1. `yarn cypress-open` in ui/apps/platform
    * compliance/complianceDashboard.test.js
    * compliance/complianceList.test.js deleted obscure negative assertion that distinguishes cluster from control side panel, but only a few controls render ControlRelatedResource elements.
